### PR TITLE
[BlueOyster] Single-domain learning

### LINKED ISSUE #1580: Weirdness in network/domains mapping
Even after #1576, we still have to reconcile the fact that `domains` were designed to be a set of identifiers for different learning contexts, while `network` is intended to be an identifier for a single "smart contracts" context.

This problem is specially problematic for StakeHolder as its configuration inherits from `CharacterConfiguration`, which has `domains`, but StakeHolder should have a single `network` instead.

As a consequence, we end up with weird mappings like:
`network=list(self.domains)[0]`  (see e.g., nucypher/config/node.py after #1576)
or the other way around:
`domains={self.network}`  (see e.g., nucypher/cli/commands/stake.py after #1576)

### LINKED ISSUE #1869: Test with a hard-coded, versioned node metadata bytestring
Imagine this situation:
1. Let's assume we start with our current LearningLoop version, which is version 1
2. Now I introduce a change in the LearningLoop that affects the metadata and that's not backwards-compatible.
3. In principle, I should also increment the LL version to version 2, but I forget about that.
4. If necessary, I adapt current tests to work with my new changes and move on. All tests pass.
5. The changes are merged introduced in a new release.
6. Suddenly we have nodes in the network with the same LL version but different metadata format.

@jMyles @KPrasch Is this a plausible scenario? If so, I propose the following:

- Introduce a new test specific to version 1. This test will include a hard-coded node metadata bytestring, which will be validated by test ursulas. The purpose of this test is to fail if a change in the codebase affects node metadata and the programmer forgets to increment the version.
- When we need to upgrade  to version 2, we need to produce a new bytestring.

As an alternative, we can introduce a test that disassemblies node metadata bytestrings of current test ursulas, and tries to validate the format against the current specification (currently, version 1). Perhaps this is even easier. We'd have to be very clear that this test cannot be changed, and that it fails is an indication that the format has changed.

### LINKED ISSUE #1871: Increment LearningLoop version and see what happens
Just because.

### LINKED ISSUE #2236: Relocate `nucypher.datastore.keypairs.py` to `nucypher.crypto`
For some reason, keypairs' logic was left in the datastore module after the last reorganization. It should live on `nucypher.crypto`.

### PR DESCRIPTION
**Pull request by cygnusv**
As agreed, for the moment we're done with multiple domains. This PR makes the necessary changes to use a single domain everywhere. 

**Closes:** 
* Closes #1871: Bumped Learning Loop version from 1 to 2.
* Closes #1869: See new tests in `tests/integration/learning/test_learning_versions.py`.
* Closes #2236: Relocate `keypairs.py` from `datastore` to `crypto`.
* Closes #1580: For the moment, networks == domains

### CODE REVIEW

File: nucypher/config/node.py [hunk 1]
```diff
@@ -55,7 +55,7 @@ class CharacterConfiguration(BaseConfiguration):
     'Sideways Engagement' of Character classes; a reflection of input parameters.
     """
 
-    VERSION = 1  # bump when static payload scheme changes
+    VERSION = 2  # bump when static payload scheme changes
```
Comment (derekpierre) [inline nucypher/config/node.py:58]: 👍

File: nucypher/network/nodes.py [hunk 2]
```diff
@@ -985,9 +979,19 @@ class UnbondedWorker(InvalidNode):
...
+        """Raised when deserializing a Character from a unexpected and incompatible version."""
+
+    class IsFromTheFuture(UnexpectedVersion):
         """Raised when deserializing a Character from a future version."""
 
+    class AreYouFromThePast(UnexpectedVersion):
+        """Raised when deserializing a Character from a previous, now unsupported version."""
+
+    unknown_version_message = "{} purported to be of version {}, but we're version {}."
+    really_unknown_version_message = "Unable to glean address from node that perhaps purported to be version {}. " \
```
Comment (derekpierre) [inline nucypher/network/nodes.py:992]: "perhaps purported" feels redundant.

```suggestion
    really_unknown_version_message = "Unable to glean address from node that purported to be version {}. " \
```

**Review by derekpierre — APPROVED** [review]
🎸 - just one suggestion.

File: examples/finnegans_wake_demo/finnegans-wake-demo.py [hunk 3]
```diff
@@ -75,7 +75,7 @@
 ######################################
 
 ALICE = Alice(network_middleware=RestMiddleware(),
-              domains={TEMPORARY_DOMAIN},
+              domain=TEMPORARY_DOMAIN,
```
Comment (KPrasch) [inline examples/finnegans_wake_demo/finnegans-wake-demo.py:78]: Ahhh - so refreshing.

File: nucypher/blockchain/eth/actors.py [hunk 4]
```diff
@@ -107,8 +107,8 @@ def __init__(self, registry: BaseContractRegistry, domains=None, checksum_addres
             self.checksum_address = checksum_address  # type: str
 
         self.registry = registry
-        if domains:  # StakeHolder config inherits from character config, which has 'domains' - #1580
-            self.network = list(domains)[0]
+        if domain:  # StakeHolder config inherits from character config, which has 'domains' - See #1580
```
Comment (KPrasch) [inline nucypher/blockchain/eth/actors.py:110]: Just curious - why is `domain` optional here?

File: nucypher/characters/lawful.py [hunk 5]
```diff
@@ -1209,7 +1210,7 @@ def run(self,
         # if learning:  # TODO: Include learning startup here with the rest of the services?
```
Comment (KPrasch) [inline nucypher/characters/lawful.py:1210]: @jMyles - May I draw your attention here :point_up_2:

**Review by KPrasch — APPROVED** [review]
A comment and a question but no RFCs from me.  Looks good!

File: nucypher/characters/lawful.py (hunk 5 above)
Comment (jMyles) [reply nucypher/characters/lawful.py:1210]: So drawn.  How can I help?

Comment (KPrasch) [reply nucypher/characters/lawful.py:1210]: The TODO says it all

File: nucypher/characters/lawful.py [hunk 6]
```diff
@@ -1304,18 +1305,16 @@ def __bytes__(self):
...
 
-        domains = {domain.encode('utf-8') for domain in self.serving_domains}
         as_bytes = bytes().join((version,
                                  self.canonical_public_address,
-                                 bytes(VariableLengthBytestring.bundle(domains)),
+                                 bytes(VariableLengthBytestring(self.serving_domain.encode('utf-8'))),
                                  self.timestamp_bytes(),
                                  bytes(self._interface_signature),
-                                 bytes(decentralized_identity_evidence),
+                                 bytes(VariableLengthBytestring(self.decentralized_identity_evidence)),  # FIXME: Fixed length doesn't work with federated
```
Comment (jMyles) [inline nucypher/characters/lawful.py:1317]: ...why?  This isn't fixed length?

Comment (cygnusv) [reply nucypher/characters/lawful.py:1317]: It is but not for federated, where it's a constant sorrow constant.

File: nucypher/network/nodes.py (hunk 2 above)
Comment (cygnusv) [reply nucypher/network/nodes.py]: ✅

**Pull request merged**